Support for assume directive : Parse & AST modules

SunilKuravinakop opened the pull request:
Support for assume directive : Parse and AST modules

Commit:
Support for "#pragma omp assume" Directive in Parse and AST.

  Changes to be committed:
 	modified:   clang/include/clang/Parse/Parser.h
 	modified:   clang/include/clang/Sema/Sema.h
 	modified:   clang/lib/Basic/OpenMPKinds.cpp
 	modified:   clang/lib/Parse/ParseOpenMP.cpp
 	modified:   clang/lib/Sema/SemaOpenMP.cpp
 	modified:   llvm/include/llvm/Frontend/OpenMP/OMP.td

Commit:
Merge branch 'main' into assume

Commit:
Adding test for checking -ast-dump.
  Changes to be committed:
 	new file:   clang/test/OpenMP/assume_ast.cpp

Commit:
Merge branch 'main' into assume

Commit:
Merge branch 'main' into assume

Commit:
Merge branch 'main' into assume

Commit:
Merge branch 'main' into assume

Commit:
Correction for git-clang-format errors.
  Changes to be committed:
 	modified:   clang/lib/Parse/ParseOpenMP.cpp

alexey-bataev commented on clang/lib/Basic/OpenMPKinds.cpp:
```diff
@@ -760,6 +760,9 @@ void clang::getOpenMPCaptureRegions(
       CaptureRegions.push_back(OMPD_parallel);
       break;
+    case OMPD_assume:
+      CaptureRegions.push_back(OMPD_assume);
```
assume ends up with a simple assume intrinsic, right? Most probably non need to emit captured function, so it should be OMPD_unknown

alexey-bataev commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -1759,8 +1759,35 @@ void Parser::ParseOpenMPAssumesDirective(OpenMPDirectiveKind DKind,
 
+  StmtResult AssociatedStmt;
+
+  // Fix the scope for assume.
```
What's wrong with the scope?

SunilKuravinakop replied:
The error was in the wording in the comment. I have corrected it to "Begin marking the scope for assume."

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -7323,6 +7324,69 @@ void SemaOpenMP::ActOnFinishedFunctionDefinitionInOpenMPAssumeScope(Decl *D) {
+
+  if (AssociatedStmt) {
+    // Add the AssumeAttr to the Directive associated with the Assume Directive.
+    OMPAssumeStmtVisitor Visitor(&OMPAssumeScoped);
+    Visitor.Visit(AssociatedStmt);
```
Why need to do this?

SunilKuravinakop replied:
This is the code which ensures that :
`-OMPAssumeAttr 0x15b07b0 <line:21:15> "omp_no_openmp"`
is added to the statement associated with the assume directive. The test case attached with this upload checks for the OMPAssumeAttr.

alexey-bataev replied:
Why do you need the visitor?

SunilKuravinakop replied:
If the example is :
 ```
// assume is for the "simd" region
  #pragma omp assume no_openmp
  #pragma omp simd
  for (int i = 0; i < N; ++i){
    A[i] += B[i];
  }
```
then in -ast-dump for:
```
    |-OMPSimdDirective 0x15cc6a0 <line:22:3, col:19>
    |  `-CapturedStmt 0x15b0d98 <line:23:3, line:25:3>
    |    `-CapturedDecl 0x15b0960 <<invalid sloc>> <invalid sloc>
    |     |
```
```
    |     |-OMPAssumeAttr 0x15b07b0 <line:21:15> "omp_no_openmp"
```
needs to be added. Also, according to the spec :
`The scope of the assume directive is the code executed in the corresponding region or in any region that is nested in the corresponding region. `
Is there a better way to add this other than having
`OMPAssumeStmtVisitor::VisitCapturedStmt( )`

alexey-bataev replied:
Why it is not enough to mark just the outer region? The fact that it is applied to nested regions can be hanf]dled in the codegen

Commit:
Changes to take care of review comments:
1) assume ends up with a simple assume intrinsic, right? Most probably non need to emit captured function, so it should be OMPD_unknown
2) "Fix the scope for assume" reworded to "Begin marking the scope for assume."
 	modified:   clang/lib/Basic/OpenMPKinds.cpp
 	modified:   clang/lib/Parse/ParseOpenMP.cpp
 	modified:   clang/test/OpenMP/assume_ast.cpp

alexey-bataev commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -1759,8 +1759,35 @@ void Parser::ParseOpenMPAssumesDirective(OpenMPDirectiveKind DKind,
+  StmtResult AssociatedStmt;
+
+  // Begin marking the scope for assume.
+  if (DKind == llvm::omp::Directive::OMPD_assume) {
+
+    if (Tok.getKind() == clang::tok::annot_pragma_openmp_end)
+      ConsumeAnyToken();
+
+    DeclarationNameInfo DirName;
+    Actions.OpenMP().StartOpenMPDSABlock(DKind, DirName, Actions.getCurScope(),
+                                         Loc);
+  }
```
Introduce RAAI class for this and use it instead

cor3ntin commented:
Can we get a more specific title for this PR? thanks!

kparzysz commented on clang/lib/Basic/OpenMPKinds.cpp:
```diff
@@ -760,6 +760,9 @@ void clang::getOpenMPCaptureRegions(
     case OMPD_parallel:
       CaptureRegions.push_back(OMPD_parallel);
       break;
+    case OMPD_assume:
+      CaptureRegions.push_back(OMPD_unknown);
+      break;
```
Please don't push OMPD_unknown here, instead add OMPD_assume to the directives at lines 790+.

kparzysz replied:
Also, please add OMPD_assume to `isOpenMPCapturingDirective`.

alexey-bataev replied:
Why? We usually add OMPD_unknown to the directives that do nit require outlining to a standalone function, here should be the same behavior.

kparzysz replied:
This is what happens with the directives at lines 790+ (before these changes).  We don't want to be adding OMPD_unknown as we go, because it can end up intermingled with other directives (in case of combined directives).  I understand that OMPD_assumes may be standalone, but we should treat it the same as the combinable ones for consistency.

alexey-bataev replied:
It may affect codegen + may add some memory overhead

kparzysz replied:
The final output from this (global) function will be the same, so it shouldn't affect codegen.

Commit:
1) Introducing RAII class for "#pragma omp assume" in ParseOpenMP.cpp
2) returning OMPD_assume in clang/lib/Basic/OpenMPKinds.cpp.

  Changes to be committed:
 	modified:   clang/include/clang/Parse/Parser.h
 	modified:   clang/lib/Basic/OpenMPKinds.cpp
 	modified:   clang/lib/Parse/ParseOpenMP.cpp

Commit:
Merge branch 'main' into assume

SunilKuravinakop commented:
In clang/lib/Sema/SemaOpenMP.cpp I have retained the marking of the inner regions of the associated statements with AssumeAttr. This will make the work in Codegen easier.

SunilKuravinakop commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -488,6 +488,49 @@ class Parser : public CodeCompletionHandler {
   StmtResult handleExprStmt(ExprResult E, ParsedStmtContext StmtCtx);
 
+  class AssumeParseAssociatedStmtRAII {
+    Parser *parent;
```
Since I am creating instance of `AssumeParseAssociatedStmtRAII` in `Parser::ParseOpenMPAssumesDirective()` I have to store Parent as a pointer. The creation of the instance is done as :
`AssumeParseAssociatedStmtRAII AssumeParseAssocRAII(this, Loc, DKind);`

alexey-bataev replied:
Pass it as *this

Commit:
Changing "parent" variable name to "Parent".
 	modified:   clang/include/clang/Parse/Parser.h

Commit:
Merge branch 'main' into assume

Commit:
1) Using "*this" as parameter.
2) Changing "Parent" to "&Parent".

 	modified:   clang/include/clang/Parse/Parser.h
 	modified:   clang/lib/Parse/ParseOpenMP.cpp

Commit:
Merge branch 'main' into assume

alexey-bataev commented:
> In clang/lib/Sema/SemaOpenMP.cpp I have retained the marking of the inner regions of the associated statements with AssumeAttr. This will make the work in Codegen easier.

How it will make the codegen easier? It is better to try to avoid post-building attribute assignment, if possible. If possible to add attribute on the region build, better to do it.

Commit:
According to OpenMP Spec, "The scope of the assume directive is the code
executed in the corresponding region or in any region that is nested in the
corresponding region". Moving the marking of the nested regions into
Codegen. Marking the outer Associated Statement provides sufficient
information.

  Changes to be committed:
 	modified:   clang/lib/Sema/SemaOpenMP.cpp

Commit:
Merge branch 'main' into assume

SunilKuravinakop committed:
Merge branch 'main' into assume

SunilKuravinakop commented:
Changes to `omp assume` was done by Julian Brown from AMD on August 5th, https://github.com/llvm/llvm-project/pull/92731 
git log:
```
commit a42e515e3a9f3bb4e44389c097b89104d95b9b29
Author: Julian Brown <julian.brown@amd.com>
Date:   Mon Aug 5 12:37:07 2024 +0100
```
Hence closing this pull request.
